cmd/cosign/cli/verify/verify_blob_attestation.go

```diff
@@ -0,0 +1,167 @@
+	if c.SignaturePath == "" {
+		return fmt.Errorf("please specify path to the base64 encoded DSSE envelope signature via --signature")
+	}
+
+	// TODO: Add support for security keys and keyless signing
+	if !options.OneOf(c.KeyRef) {
+		return &options.PubKeyParseError{}
+	}
+
+	var err error
+	co := &cosign.CheckOpts{}
+
+	if c.CheckClaims {
+		co.ClaimVerifier = cosign.IntotoSubjectClaimVerifier
+	}
+
+	keyRef := c.KeyRef
+
+	// TODO: keyless signing
+	co.SigVerifier, err = sigs.PublicKeyFromKeyRef(ctx, keyRef)
+	if err != nil {
+		return fmt.Errorf("loading public key: %w", err)
+	}
+	pkcs11Key, ok := co.SigVerifier.(*pkcs11key.Key)
+	if ok {
+		defer pkcs11Key.Close()
+	}
+
+	// Read the signature and decode it (it should be base64-encoded)
+	encodedSig, err := os.ReadFile(filepath.Clean(c.SignaturePath))
+	if err != nil {
+		return fmt.Errorf("reading %s: %w", c.SignaturePath, err)
+	}
+	decodedSig, err := base64.StdEncoding.DecodeString(string(encodedSig))
+	if err != nil {
+		return fmt.Errorf("decoding signature: %w", err)
+	}
+
+	// Verify the signature on the attestation against the provided public key
+	env := ssldsse.Envelope{}
+	if err := json.Unmarshal(decodedSig, &env); err != nil {
+		return fmt.Errorf("marshaling envelope: %w", err)
+	}
+
+	if env.PayloadType != types.IntotoPayloadType {
+		return cosign.NewVerificationError("invalid payloadType %s on envelope. Expected %s", env.PayloadType, types.IntotoPayloadType)
+	}
+	dssev, err := ssldsse.NewEnvelopeVerifier(&dsse.VerifierAdapter{SignatureVerifier: co.SigVerifier})
+	if err != nil {
+		return fmt.Errorf("new envelope verifier: %w", err)
+	}
+	if _, err := dssev.Verify(&env); err != nil {
+		return fmt.Errorf("dsse verify: %w", err)
+	}
+
+	// Verify the attestation is for the provided blob and the predicate type
+	if err := verifyBlobAttestation(env, artifactPath, c.PredicateType); err != nil {
+		return err
+	}
+
+	fmt.Fprintln(os.Stderr, "Verified OK")
+	return nil
+}
+
+func verifyBlobAttestation(env ssldsse.Envelope, blobPath, predicateType string) error {
+	artifact, err := os.ReadFile(blobPath)
+	if err != nil {
+		return fmt.Errorf("reading %s: %w", blobPath, err)
+	}
+
+	// Get the actual digest of the blob
+	digest, _, err := signature.ComputeDigestForSigning(bytes.NewReader(artifact), crypto.SHA256, []crypto.Hash{crypto.SHA256, crypto.SHA384})
+	if err != nil {
+		return err
+	}
+	actualDigest := strings.ToLower(hex.EncodeToString(digest))
+
+	// Get the expected digest from the attestation
+	decodedPredicate, err := base64.StdEncoding.DecodeString(env.Payload)
+	if err != nil {
+		return fmt.Errorf("decoding dsse payload: %w", err)
+	}
+	var statement in_toto.Statement
+	if err := json.Unmarshal(decodedPredicate, &statement); err != nil {
+		return fmt.Errorf("decoding predicate: %w", err)
+	}
+
+	// Compare the actual and expected
+	if statement.Subject == nil {
+		return fmt.Errorf("no subject in intoto statement")
+	}
+	if len(statement.Subject) != 1 {
```

Comment:
should this definitely be a requirement? if the attestation signed over say, a jar file and a pom file, the attestation should verify against the jar file, right?

Reply:
it's sad that `IntotoSubjectClaimVerifier` logic needs refactoring to be reused..

Reply:
Ah I added this in because `attest-blob` only allows one subject at the moment. I can remove this requirement in a follow-up PR though!